File: 2026-01-January-GFG-POTD/January 20 - Implement UNDO & REDO/Solution.py
class Solution:
    def __init__(self):
        self.document = []
        self.redoStack = []

    def append(self, x):
        self.document.append(x)
        self.redoStack.clear()  # clear redo history

    def undo(self):
        if self.document:
            self.redoStack.append(self.document.pop())

    def redo(self):
        if self.redoStack:
            self.document.append(self.redoStack.pop())

    def read(self):
        return ''.join(self.document)

File: 2026-01-January-GFG-POTD/January 20 - Implement UNDO & REDO/Solution.java
class Solution {
    private StringBuilder document = new StringBuilder();
    private Stack<Character> redoStack = new Stack<>();

    public void append(char x) {
        document.append(x);
        redoStack.clear(); // new change invalidates redo history
    }

    public void undo() {
        if (document.length() > 0) {
            char removed = document.charAt(document.length() - 1);
            document.deleteCharAt(document.length() - 1);
            redoStack.push(removed);
        }
    }

    public void redo() {
        if (!redoStack.isEmpty()) {
            document.append(redoStack.pop());
        }
    }

    public String read() {
        return document.toString();
    }
}
